core/src/main/java/io/grpc/internal/ServerCallImpl.java

```diff
@@ -55,6 +63,8 @@
   private boolean sendHeadersCalled;
   private boolean closeCalled;
   private Compressor compressor;
+  private boolean messageSent;
+  private boolean internalClosed;
```

Comment:
I'm pretty sure we're [relying on similar behavior client-side](https://github.com/grpc/grpc-java/blob/v1.4.0/core/src/main/java/io/grpc/internal/ClientCallImpl.java#L395). I would believe it may be slightly broken, but I'd be in favor of trying to make it a guarantee of the stream.